### PR TITLE
Scala 3 (Dotty) support

### PR DESCRIPTION
**Pull request by larsrh**
Collaborative PR: @alexandru @Avasil @larsrh 

<hr>

Prequel PRs:

* [x] #1326 
* [x] #1327 
* [x] #1328 
* [x] #1329
* [x] #1331
* [x] #1344

### CODE REVIEW

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/GroupByOperator.scala
Comment (larsrh) [inline monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/GroupByOperator.scala:54]: Are you sure you want to remove all those `@tailrec` annotations? We could rewrite the functions to make them tail recursive.

Comment (Avasil) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/GroupByOperator.scala:54]: They are still tail recursive, it just can't be verified with the annotation, right?

It would be preferable to keep the annotation but I'm not sure how to do that in cases like:

```scala
  def unsafeSubscribeFn(subscriber: Subscriber[A]): Cancelable = {
    import subscriber.scheduler
    val state = stateRef.get()

    if (state.errorThrown != null) {
      subscriber.onError(state.errorThrown)
      Cancelable.empty
    } else if (state.isDone) {
      Observable
        .now(state.cached)
        .unsafeSubscribeFn(subscriber)
    } else {
      val c = ConnectableSubscriber(subscriber)
      val newState = state.addNewSubscriber(c)

      if (stateRef.compareAndSet(state, newState)) {
        c.pushFirst(state.cached)
        val connecting = c.connect()

        val cancelable = Cancelable { () =>
          removeSubscriber(c)
        }
        connecting.syncOnStopOrFailure(_ => cancelable.cancel())
        cancelable
      } else {
        // retry
        unsafeSubscribeFn(subscriber)
      }
    }
  }
```

Comment (larsrh) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/GroupByOperator.scala:54]: It's true, but in case something changes in the future, the compiler won't tell us if it's still tail recursive. I refactored this in #1326 by extracting the core loop into an inner method that I've annotated accordingly. However, I don't have any strong opinions on this.

Comment (Avasil) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/GroupByOperator.scala:54]: Ahh, I was too hasty - I've corrected these method in the latest commit

File: monix-reactive/shared/src/test/scala/monix/reactive/internal/operators/PublishSelectorSuite.scala
Comment (Avasil) [inline monix-reactive/shared/src/test/scala/monix/reactive/internal/operators/PublishSelectorSuite.scala:35]: There is this error (doesn't happen for `Int` ) without type ascription:

```
|ambiguous implicit arguments: both object IntIsIntegral in object Numeric and object DoubleIsFractional in object Numeric match type Numeric[B] of parameter B of method sumL in class Observable
```

didn't have time to investigate why

**Comment by larsrh** [comment]
There a still four compile errors in the doctest `reactiveJVM`:

```
[error] -- Error: /home/lars/proj/monix/monix/monix-reactive/jvm/target/scala-3.0.0-M1/src_managed/test/monix/reactive/ObservableDoctest.scala:446:11 
[error] 446 |    ).merge
[error]     |           ^
[error]     |no implicit argument of type monix.reactive.OverflowStrategy[B] was found for parameter os of method merge in class Observable
```

I find that strange; I thought the type ascription to the implicit argument of `merge` fixed these.

**Comment by Avasil** [comment]
> There a still four compile errors in the doctest `reactiveJVM`:
> 
> ```
> [error] -- Error: /home/lars/proj/monix/monix/monix-reactive/jvm/target/scala-3.0.0-M1/src_managed/test/monix/reactive/ObservableDoctest.scala:446:11 
> [error] 446 |    ).merge
> [error]     |           ^
> [error]     |no implicit argument of type monix.reactive.OverflowStrategy[B] was found for parameter os of method merge in class Observable
> ```
> 
> I find that strange; I thought the type ascription to the implicit argument of `merge` fixed these.

I've opened an issue about it https://github.com/lampepfl/dotty/issues/10497

**Comment by larsrh** [comment]
> I've opened an issue about it

Ah right, I forgot that existed :grin:

**Comment by Avasil** [comment]
@larsrh I have finally spent some time on it and the good news is that almost all tests pass! I think the only issue left is `Observable#merge`...

File: monix-execution/shared/src/main/scala_2.13+/monix/execution/compat.scala
Comment (Avasil) [inline monix-execution/shared/src/main/scala_2.13+/monix/execution/compat.scala:34]: I think `inline` made some difference in some `bufferSliding` benchmarks ( https://github.com/monix/monix/blob/series/4.x/benchmarks/shared/src/main/scala/monix/benchmarks/ChunkedEvalFilterMapSumBenchmark.scala#L123 ) but I'm not 100% sure

Comment (alexandru) [reply monix-execution/shared/src/main/scala_2.13+/monix/execution/compat.scala:34]: Maybe. But it's risky, and I'd rather get rid of these. We can redo the benchmarks, and then add some explicit comments for that inlining, to ensure that we did our homework. 

What I want to avoid is overuse of the `inline` annotation, which in the beginning I believed that it was a good idea, and added such annotations based on gut feeling.

Comment (Avasil) [reply monix-execution/shared/src/main/scala_2.13+/monix/execution/compat.scala:34]: Makes sense, I'll just rerun the benchmark before releasing

**Comment by alexandru** [comment]
I solved `merge` by fixing the tests 🤷‍♂️ as there's nothing we can do for now, the only other options being:

1. to wait for that bug to get fixed in Scala 3, or ...
2. break compatibility

We can break compatibility in 4.0.0, but for now this will have to do.

---

I also upgraded sbt, and cleaned up all warnings, except for those weird ones that aren't reported as actual warnings and that may be the result of some plugin.

Not sure if I want to do anything else before merging.

**Review by Avasil — APPROVED** [review]

**Comment by larsrh** [comment]
Please hold back with merging until I get the chance to do one final review pass, okay?

**Comment by alexandru** [comment]
@larsrh sounds good, thanks!

File: monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala
Comment (oleg-py) [inline monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala:29]: So, Scala 3 dropped adding object prefixes to implicit scope, so the trick I proposed quite some time ago with making these options work out of the box is broken. I wonder if we can/should adapt something like technique from [here](https://blog.7mind.io/no-more-orphans.html) or simply bite the bullet and drop the "monix-execution doesn't depend on cats directly" policy.

File: monix-execution/js/src/main/scala_3.0-/monix/execution/atomic/AtomicNumber.scala
Comment (oleg-py) [inline monix-execution/js/src/main/scala_3.0-/monix/execution/atomic/AtomicNumber.scala:56]: :D

**Review by oleg-py — APPROVED** [review]
Looks good to me

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnTerminateOperator.scala
Comment (alexandru) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnTerminateOperator.scala:74]: Because the inner-method captures the current context, it can do an extra-allocation, since it's probably going to get translated into a class initialized with whatever it captures in its constructor. Moving it outside ensures it's not capturing anything, and invoking it will not create an extraneous object.

File: monix-execution/shared/src/main/scala_3.0-/monix/execution/misc/test/TestInlineMacros.scala
Comment (alexandru) [reply monix-execution/shared/src/main/scala_3.0-/monix/execution/misc/test/TestInlineMacros.scala:37]: This was probably a mistake. One of these days we should just enable Scalafmt in the build.

File: monix-eval/shared/src/main/scala/monix/eval/internal/TaskRunToFutureWithLocal.scala
Comment (alexandru) [reply monix-eval/shared/src/main/scala/monix/eval/internal/TaskRunToFutureWithLocal.scala:71]: Because there we need an `Any => Task[Any]` and it doesn't match, compiler inferring `Nothing`, can't remember where.

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnTerminateOperator.scala
Comment (oleg-py) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnTerminateOperator.scala:74]: Compiler has never done that. The desugaring of inner methods is actually pretty much the same as hand-written there. You can see it by doing `-Xprint:lambdalift` (scala2) or `-Xprint:lambdaLift` (scala 3). You can check it easily on Scastie, [e.g. this](https://scastie.scala-lang.org/qwOq9riDRQWSHi6zVJDsRQ)

File: monix-eval/shared/src/main/scala/monix/eval/internal/TaskRunToFutureWithLocal.scala
Comment (oleg-py) [reply monix-eval/shared/src/main/scala/monix/eval/internal/TaskRunToFutureWithLocal.scala:71]: Oh yeah dotty inferring Nothing/Any changed in way I don't quite understand but was annoyed by so I think I get what you're talking about

File: monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala
Comment (alexandru) [reply monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala:29]: I don't know what to do next. I don't like the "no more orphans" solution, as it looks like something that can break in the future. I also noticed that Scala 3 started to give advice on imports, not sure how reliable it is, but it looks good. The Scala 3 compiler actually suggested that import in the error report.

Another possibility would be to have another sub-project called `monix-execution-instances` (also a dependency of `monix-catnap`) and have people do ...

```scala
import monix.execution.instances._
```

And hopefully the Scala 3 compiler will suggest it.

Comment (alexandru) [reply monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala:29]: What I mean is, if we want an optional dependency on Cats, and if the Scala 3 compiler (aka the future) can suggest the import, then maybe such orphans are fine.

Comment (oleg-py) [reply monix-eval/jvm/src/test/scala/monix/eval/TaskLocalJVMSuite.scala:29]: Yeah I see what you mean. I'd put those into catnap to not breed submodules.

File: monix-eval/shared/src/main/scala/monix/eval/internal/TaskRunToFutureWithLocal.scala
Comment (larsrh) [reply monix-eval/shared/src/main/scala/monix/eval/internal/TaskRunToFutureWithLocal.scala:71]: The problem here is that the type of `bindNext` contains a skolem, whose handling in Scala 2 is unsound. Scala 3 tightened this.

File: monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/Atomic.scala
Comment (larsrh) [inline monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/Atomic.scala:40]: same

File: monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/Atomic.scala
Comment (larsrh) [inline monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/Atomic.scala:28]: Macro in Scala 2, make `inline`?

File: monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/Atomic.scala
Comment (larsrh) [inline monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/Atomic.scala:46]: same

File: monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/AtomicNumber.scala
Comment (larsrh) [inline monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/AtomicNumber.scala:31]: same

File: monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/AtomicNumber.scala
Comment (larsrh) [inline monix-execution/jvm/src/main/scala_3.0/monix/execution/atomic/AtomicNumber.scala:37]: same

File: monix-execution/shared/src/main/scala/monix/execution/Scheduler.scala
Comment (larsrh) [inline monix-execution/shared/src/main/scala/monix/execution/Scheduler.scala:412]: Is this accidental or on purpose?

File: monix-execution/shared/src/main/scala/monix/execution/schedulers/SchedulerService.scala
Comment (larsrh) [inline monix-execution/shared/src/main/scala/monix/execution/schedulers/SchedulerService.scala:121]: same here

File: monix-execution/shared/src/main/scala_3.0/monix/execution/misc/Local.scala
Comment (larsrh) [inline monix-execution/shared/src/main/scala_3.0/monix/execution/misc/Local.scala:138]: used to be a macro in Scala 2

File: monix-java/src/main/scala/monix/java8/eval/package.scala
Comment (larsrh) [inline monix-java/src/main/scala/monix/java8/eval/package.scala:33]: Why is this undeprecated?

File: monix-java/src/main/scala/monix/java8/execution/package.scala
Comment (larsrh) [inline monix-java/src/main/scala/monix/java8/execution/package.scala:34]: Why is this undeprecated?

File: monix-java/src/main/scala/monix/java8/execution/package.scala
Comment (larsrh) [inline monix-java/src/main/scala/monix/java8/execution/package.scala:68]: same

**Review by larsrh — COMMENTED** [review]
I noted a few issues where there are inconsistencies between Scala 2 and 3 and (possibly?) unrelated changes that I don't understand. But those aren't blockers. Good work, everyone :muscle:

File: monix-execution/js/src/main/scala_3.0/monix/execution/atomic/AtomicNumber.scala
Comment (alexandru) [reply monix-execution/js/src/main/scala_3.0/monix/execution/atomic/AtomicNumber.scala:31]: We have abused macros in `Atomic` and it's time to take a step back. I don't think those aliases being defined as macros have any value. Macros help in those `transform` methods for inlining the given function parameter in what is a hot loop, otherwise the JVM is perfectly capable of inlining such method calls by itself.

That said maybe we should have feature parity and then get rid of those macros in 4.0.0.

So I'm going to revert this change, to have consistency.

File: monix-execution/shared/src/main/scala/monix/execution/Scheduler.scala
Comment (alexandru) [reply monix-execution/shared/src/main/scala/monix/execution/Scheduler.scala:412]: It's on purpose. The Scala 3 compiler ends up triggering that warning even when those extensions methods aren't used, don't know why.

Deprecating just the methods is enough IMO. Who knows, we might want to add other extensions in the future.

File: monix-java/src/main/scala/monix/java8/eval/package.scala
Comment (alexandru) [reply monix-java/src/main/scala/monix/java8/eval/package.scala:33]: For the same reason. Scala 3 compiler triggers that warning even when not used. Don't know if this is a bug or what changed in Scala 3.

I have a zero tolerance for warnings :)

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnTerminateOperator.scala
Comment (alexandru) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnTerminateOperator.scala:74]: Hmm, my knowledge there may be outdated, goes to show that we should always check and re-check our assumptions.

Will revert the code change.

File: monix-execution/shared/src/main/scala_3.0/monix/execution/misc/Local.scala
Comment (alexandru) [reply monix-execution/shared/src/main/scala_3.0/monix/execution/misc/Local.scala:138]: Nice catch.

File: monix-java/src/main/scala/monix/java8/execution/package.scala
Comment (alexandru) [reply monix-java/src/main/scala/monix/java8/execution/package.scala:68]: Same as the others — Scala 3 generates warnings, even if the extensions don't get used, which is why I prefer to have the deprecation warnings only on the methods themselves.

File: monix-java/src/main/scala/monix/java8/execution/package.scala
Comment (alexandru) [reply monix-java/src/main/scala/monix/java8/execution/package.scala:34]: Same as the others — Scala 3 generates warnings, even if the extensions don't get used, which is why I prefer to have the deprecation warnings only on the methods themselves.

File: monix-execution/shared/src/main/scala_3.0/monix/execution/misc/Local.scala
Comment (alexandru) [reply monix-execution/shared/src/main/scala_3.0/monix/execution/misc/Local.scala]: Solved it.

**Pull request merged**
